lib/log.h

```diff
@@ -0,0 +1,140 @@
+ * compiler option.
+ *   -DDEBUG or -DDEBUG=1 enables the debug logging via T_DBG().
+ *   -DDEBUG=2 - same as above, but also enables T_DBG2().
+ *   -DDEBUG=3 - same as above plus T_DBG3().
+ *   ...etc
+ * Currently there are only 3 levels:
+ *   1 [USER]    - information required to understand system behavior under
+ *                 some load, only key events (e.g. new connections) should
+ *                 be logged. The events could not be logged on normal level,
+ *                 because we expect too many such events. This level should
+ *                 be used only for events interesting to common system
+ *                 administrator;
+ *   2 [SUPPORT] - key events at lower (component) levels (e.g. TDB or SS).
+ *                 Only events required for technical support should be logged
+ *                 on this level;
+ *   3 [DEVELOP] - verbose logging, used for engineer debugging internal
+ *                 algorithms and so on. Typically for single slow connection
+ *                 cases.
+ */
+#define __BNR		"[tempesta " BANNER "] "
+#define __T_DBG1(...) 	pr_debug(__BNR "  " __VA_ARGS__)
+#define __T_DBG2(...) 	pr_debug(__BNR "    " __VA_ARGS__)
+#define __T_DBG3(...)	pr_debug(__BNR "      " __VA_ARGS__)
+
+#if defined(DEBUG) && (DEBUG >= 1)
+#define T_DBG(...) 	__T_DBG1(__VA_ARGS__)
+#else
+#define T_DBG(...)
+#endif
+
+#if defined(DEBUG) && (DEBUG >= 2)
+#define T_DBG2(...)	__T_DBG2(__VA_ARGS__)
+#else
+#define T_DBG2(...)
+#endif
+
+#if defined(DEBUG) && (DEBUG == 3)
+#define T_DBG3(...)	__T_DBG3(__VA_ARGS__)
+
+#define T_DBG3_BUF(str, buf, len)					\
+do {									\
+	T_DBG3("dump '" str "' (ptr=%pk, %lu bytes):\n",		\
+	       buf, (size_t)(len));					\
+	print_hex_dump_bytes(__BNR "        ", DUMP_PREFIX_OFFSET, buf, len);\
+} while (0)
+
+#define T_DBG3_SL(str, sglist, sgn, off, len)				\
+do {									\
+	int __o = 0, __i, __n = (int)(len);				\
+	struct scatterlist *__s = NULL;					\
+	T_DBG3(str " (sgn=%u sglist=%pK len=%lu):\n", sgn, sglist, len);\
+	for_each_sg(sglist, __s, sgn, __i) {				\
+		if (!__n)						\
+			break;						\
+		if (off >= __o + __s->length) {				\
+			__o += __s->length;				\
+		} else {						\
+			int __d = off - __o;				\
+			int __bytes = min_t(int, __n, __s->length - __d); \
+			T_DBG3("=> segment %d, ptr=%pK len=%u\n",	\
+			       __i, __s, __s->length);			\
+			T_DBG3_BUF("data", sg_virt(__s) + __d, __bytes); \
+			__o = off;					\
+			__n -= __bytes;					\
+		}							\
+	}								\
+} while (0)
+
+#define __CALLSTACK_MSG(...)						\
+do {									\
+	printk(__VA_ARGS__);						\
+	__WARN();							\
+} while (0)
+
+#define T_ERR(...)	__CALLSTACK_MSG(KERN_ERR __BNR "ERROR: " __VA_ARGS__)
+#define T_WARN(...)	__CALLSTACK_MSG(KERN_WARNING __BNR		\
+					"Warning: " __VA_ARGS__)
+#define T_LOG(...)	pr_info(__BNR __VA_ARGS__)
+/* Non-limited printing. */
+#define T_ERR_NL(...)	T_ERR(__VA_ARGS__)
+#define T_WARN_NL(...)	T_WARN(__VA_ARGS__)
+#define T_LOG_NL(...)	T_LOG(__VA_ARGS__)
```

Comment:
Non limited printing is normally USER level warnings like errors or warnings  while configuration parsing. There is no need to print call stacks here. `TFW_(ERR|WARN|LOG)` messages are also used for USER-level only messages and callstacks are not really required there and it sometimes annoying to search though system log and search between callstacks while some of them are not internal errors.